src/main/java/org/breedinginsight/services/ProgramUploadService.java

```diff
@@ -0,0 +1,233 @@
+        // because by making the PUT call the client already expected an overwrite
+        programUploadDao.deleteUploads(programId, actingUser.getId(), UploadType.TRAIT);
+
+        // do not autopopulate fields, that will be done on actual trait creation
+        BatchUploadEntity uploadEntity = BatchUploadEntity.builder()
+                .type(UploadType.TRAIT)
+                .programId(programId)
+                .userId(actingUser.getId())
+                .data(JSONB.valueOf(json))
+                .createdBy(actingUser.getId())
+                .updatedBy(actingUser.getId())
+                .build();
+
+        // Insert and update
+        programUploadDao.insert(uploadEntity);
+        return programUploadDao.getUploadById(uploadEntity.getId()).get();
+    }
+
+    private void checkRequiredTraitFields(Trait trait) throws UnprocessableEntityException {
+
+        Method method = trait.getMethod();
+        Scale scale = trait.getScale();
+
+        if (method == null) {
+            throw new UnprocessableEntityException("Missing method");
+        }
+
+        if (scale == null) {
+            throw new UnprocessableEntityException("Missing scale");
+        }
+
+        if (isBlank(trait.getTraitName())) {
+            throw new UnprocessableEntityException("Missing trait name");
+        }
+        if (isBlank(trait.getDescription())) {
+            throw new UnprocessableEntityException("Missing trait description");
+        }
+        if (trait.getProgramObservationLevel() == null || isBlank(trait.getProgramObservationLevel().getName())) {
+            throw new UnprocessableEntityException("Missing trait level");
+        }
+        if (isBlank(method.getMethodName())) {
+            throw new UnprocessableEntityException("Missing method name");
+        }
+        if (isBlank(method.getDescription())) {
+            throw new UnprocessableEntityException("Missing method description");
+        }
+        if (isBlank(method.getMethodClass())) {
+            throw new UnprocessableEntityException("Missing method class");
+        }
+        if (isBlank(scale.getScaleName())) {
+            throw new UnprocessableEntityException("Missing scale name");
+        }
+        if (scale.getDataType() == null) {
+            throw new UnprocessableEntityException("Missing scale type");
+        }
+    }
+
+    private boolean isBlank(String field) {
+        if (field == null || field.isBlank()) {
+            return true;
+        }
+        return false;
+    }
+
+    private void checkTraitDataConsistency(Trait trait) throws UnprocessableEntityException {
+
+        Method method = trait.getMethod();
+        Scale scale = trait.getScale();
+
+        if (method != null && method.getMethodClass().equals(Method.COMPUTATION_TYPE)) {
+            if (isBlank(method.getFormula())) {
+                throw new UnprocessableEntityException("Missing formula for Computation method");
+            }
+        }
```

Comment:
I was under the impression that the method class could be user definable and isn't a strict set like the scale class. I probably got this idea from looking at BrAPI where it says, `Method class (examples: "Measurement", "Counting", "Estimation", "Computation", etc.)`. Are the method classes going to be a strict set like the scale classes?

Reply:
Ah I see. If this is a free formed field, should we even check for the formula if they have 'Computation' as the method? Or just take it as a computed field if they send a formula with it? It might not be a big deal to manage, but I'm picturing us having to hardcode up a bunch of special cases in our UI and backend in the future if 'Computation' is a special keywork, but other method classes are ignored.

Reply:
@eawoods @timparsons What are the actual expectations for the method class? Will it be an unchangeable set or something programs can change?

Reply:
My understanding is that in the short term, this will be a pre-generated list.
In the not-too-distant-but-nine-versions-from-now future (v1.0? 1.5?), this will be configurable for each program.

As I'm going through methods and scales -- there are some scales that make sense with certain methods and not so much with others. I wonder if it would make sense to have a table where they are called out (computation -> formula; measurement -> units, ordinal scale, integer, time/duration; observation -> ordinal scale, nominal scale, integer, boolean; etc). Worth considering, although it is possible that there are only a few cases that are as clear cut as computation/formula. @timparsons, your thoughts?

Reply:
@nickpalladino I know we've talked about various program-definable properties, but I don't remember if method class was in there.  Looking at the design reference for creating/editing a trait, it looks like a pre-defined list of method classes will be needed.  I think we could start with it being system-wide, then if it is determined that it's needed at a program level, we enhance the system to support that.  @eawoods please add any thoughts you may have.

Reply:
Agreed on holding off on method/scales map. Sounds like the conclusion is: yes on an unchangeable set of methods for now.

Reply:
Thanks for the input @eawoods @timparsons. @ctucker3 since method is unchangeable for now I agree with your initial proposal to add an enum to the database to match the way it's done for scales.

Reply:
Sounds good to me!

Reply:
Discussed with @ctucker3 and @timparsons, decided to make a card for this to deal with at a later time: https://trello.com/c/ucFwGAGl